### ISSUE TITLE
yq 3.3.1 is breaking an original indentation level for list items

### ISSUE THREAD

**Issue by posquit0** [mikefarah/yq#462]
**Describe the bug**
yq 3.3.1 is breaking an original indentation level for list items

**Input Yaml**
Concise yaml document(s) (as simple as possible to show the bug)
data1.yml:
```yaml
this:
- a
- b
- c
```

**Command**
The command you ran:
```
yq merge data1.yml data2.yml
```

**Actual behavior**

```yaml
this:
  - a
  - b
  - c
```

**Expected behavior**

```yaml
this:
- a
- b
- c
```

**Additional context**

**Comment by mikefarah** [comment]
Yeah this is a result of updating to the latest https://godoc.org/gopkg.in/yaml.v3 package (which is the underlying yaml parser).

It used to have an issue the other way around (if you had indentation, it would remove it) - but many people complained and so now it does this. For what it's worth, I actually think this is better - but yeah it doesn't preserve formatting perfectly.

**Comment by posquit0** [comment]
Do you mean this behavior is the expected result? :)

**Comment by mikefarah** [comment]
No - just a known issue.

I've added it to the readme too, perhaps future version of yaml.v3 will fix it

**Comment by posquit0** [comment]
Thank you! 👍🏼

**Comment by stristr** [comment]
@mikefarah, I have a generalized AST-based solution for this problem and have built it out enough to satisfy my use case (still have some edge cases to burn down where the renderer un-chomps multiline strings and reformats inline maps and sequences, but indentation is perfect).

Would you accept a contribution, e.g. the API proposed in #1359? It is a great deal of added complexity but can be isolated.

**Comment by mikefarah** [comment]
Great deal of complexity doesn't sound great...I'd be reluctant to add anything too complex in, as it would be something I would have to support going forward. 

yq has its own AST, is it using that? Can I have a look at the code?

**Comment by stristr** [comment]
> Great deal of complexity doesn't sound great...I'd be reluctant to add anything too complex in, as it would be something I would have to support going forward.

Fair! I can elaborate a bit more about why I think it's very important for `yq` to have this feature—succinctly, `yq` is not suitable as a "Swiss army knife utility" for [large-scale changes](https://abseil.io/resources/swe-book/html/ch22.html) if using it results in large-scale changes that consist largely of whitespace. `yq` is probably the best YAML tool out there in terms of fidelity to original YAML document, but the value of its ability to mutate YAML in place is weakened by its behavior in the face of idiosyncratic YAML formatting conventions in human-authored YAML. The option to set indentation helps, but many users have shared examples of cases where indentation alone is not sufficient.

```yaml
this:
 is:
         - valid
         - yaml:
           - unfortunately: for us
```

> Can I have a look at the code?

So to be clear, I've written a postprocessor of `yq` stdout that takes `yq` input and restores most of its original formatting including indentation. It's currently in Python and based on `PyYAML` AST, which was simply quickest for the project I'm working on but not the ideal final destination for such an algorithm. I'll describe the algo at a high level rather than burden you with the code:
 - Analyze old AST with depth-first search, tracking the indentation of each node against its JSON path. There's a bit of nuance here because maps, sequences, and scalar nodes all have whitespace-sensitive concerns depending on how they're specified, but I have access to a massive regression suite of human-authored YAML and feel quite confident this analysis is sound, complete, and testable.
  - Traverse new AST depth-first, mutating output along the way to restore original formatting to preexisting JSON paths and falling back to sensible defaults for JSON paths that didn't exist before or have fundamentally changed their semantics. The tricky things here are vast due to multiline strings, comment- and whitespace-only lines, trailing whitespaces on unchanged lines, single-line string quote style, and inline maps/sequences.

This can be implemented as a "best effort" algorithm that responsibly bails with a warn log if any _values_ of the rendered YAML AST change at all, safeguarding against potential defects.

I haven't yet dug into `yq` to understand the optimal integration point. If you think `gopkg.in/yaml.v3` is the optimal integration point and this is a pass-through API, I would be quite open to that but figured I'd start here where I'd like to pass a flag like `--preserve-indent` as proposed by another user in #1359.

**Comment by ingydotnet** [comment]
> Yeah this is a result of updating to the latest https://godoc.org/gopkg.in/yaml.v3 package (which is the underlying yaml parser).
> 
> It used to have an issue the other way around (if you had indentation, it would remove it) - but many people complained and so now it does this. For what it's worth, I actually think this is better - but yeah it doesn't preserve formatting perfectly.

@mikefarah can you find the go-yaml/yaml commit where this was reversed?

For yaml/go-yaml I'd like to add input comment level to the events and nodes such that it could be preserved.

I'd also like to make the behavior configurable.

**Comment by mikefarah** [comment]
Ooh that's a blast from the past - best guess is that it's this commit: https://github.com/go-yaml/yaml/commit/ae27a744346343ea814bd6f3bdd41d8669b172d0

**Comment by jandubois** [comment]
I too would love to see this fixed! We (the Lima project) run `yamlfmt` over the output from `yqlib` mostly for this reason (and for the empty line preservation).

Fixing up the formatting with `yamlfmt` means we can't use yq colorization. I just made some change to only run `yamlfmt` when output doesn't go to the terminal, but now the colored terminal output and file output look slightly different.

Maybe I'm somewhat obsessive with my dislike for the double-indentation for sequences. 😄 

So anyways, the empty lines issue can be easily worked around by turning them into a marker like `# This is actually an empty line` before passing the YAML to `yq`, and replacing it back to empty lines on the output, so the indentation fix is the only stumbling block to removing `yamlfmt` completely.

**Comment by ingydotnet** [comment]
@jandubois Could you file an issue here: https://github.com/yaml/go-yaml/issues and mention this one?

It would be great if you could point to example Lima files and the flow you currently use and the issues involved.

**Comment by zzJinux** [comment]
@mikefarah What's the end goal here? Print sequences as in pre-3.3.1 or provide an option for choosing which indentation to use? Both are feasible with https://pkg.go.dev/go.yaml.in/yaml/v4#Encoder.CompactSeqIndent

**Comment by stristr** [comment]
> What's the end goal here?

In context, these kinds of feature requests arise not exclusively in the context of sequences, but preservation of formatting with _any_ structure where presentation may change when using `yq`.

A typical scenario is general purpose codemod tools using `yq` as the applicator. Without an option to preserve original formatting, including sequence indentation, it is quite common to produce diffs that are mostly whitespace. In the context of human review, unreadable diffs from automated refactoring are more likely to be rejected or ignored. Indentation, styles of multiline strings, trailing whitespaces, consecutive newlines, etc. are all implicated in this.

There have been a few tickets for this and similar issues. It may simplify tracking to rename and repurpose this hot issue; the general request from the community is an option to behave more like OpenRewrite—a less useful tool, but one without this particular limitation.

**Comment by mikefarah** [comment]
Didn't know that was an option @zzJinux, and could def expose that as a flag to pass down - but yeah the aim is to preserve the original formatting and white-space as much as possible - for precisely the reasons @stristr point out. This flag would apply to the whole document, and would need to be set (or detected somehow?). Could be a pragmatic step forward though

**Comment by zzJinux** [comment]
@mikefarah then should I open a separete issue for adding an option controlling the sequence indentation?

**Comment by mikefarah** [comment]
Yep, sounds good 👍🏼

**Comment by ingydotnet** [comment]
FYI, go-yaml v4 is introducing a plugin system that will  let you hook into a lot of places that tools like yq will be interested in.
Each plugin type will offer a configurable API, so that you can either configure the standard plugin to do what you want or write your own plugin to the API, for more involved things.
